File: 1386-cinema-seat-allocation/1386-cinema-seat-allocation.cpp
class Solution {
public:
    int maxNumberOfFamilies(int n, vector<vector<int>>& reservedSeats) {
        int res = 2 * n;
        unordered_map<int, int> rows;
        for (int i = 0; i < reservedSeats.size(); i++) {
            int r = reservedSeats[i][0];
            int c = reservedSeats[i][1];
            if(!rows.count(r))
            {
                rows[r] = 0;
            }
            if (c == 2 || c == 3) {
                rows[r] |= 4;
            } else if (c == 4 || c == 5) {
                rows[r] |= 4;
                rows[r] |= 2;
            } else if (c == 6 || c == 7) {
                rows[r] |= 2;
                rows[r] |= 1;
            } else if (c == 8 || c == 9) {
                rows[r] |= 1;
            }
        }
        for (auto& [r, off] : rows) {
            if (off == 7) {
                res -= 2;
            } else if (off >= 1) {
                res--;
            }
        }
        return res;
    }
};

File: 1386-cinema-seat-allocation/1386-cinema-seat-allocation.py
class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: List[List[int]]) -> int:
        res = 2 * n
        rows = {}
        for r, c in reservedSeats:
            if r not in rows:
                rows[r] = 0
            if c == 2 or c == 3:
                rows[r] |= 4
            elif c == 4 or c == 5:
                rows[r] |= 4
                rows[r] |= 2
            elif c == 6 or c == 7:
                rows[r] |= 2
                rows[r] |= 1
            elif c == 8 or c == 9:
                rows[r] |= 1
        for r, off in rows.items():
            if off == 7:
                res -= 2
            elif off >= 1:
                res -= 1
        return res